### PR TITLE
support VM IP retain for kubevirt

### PR DESCRIPTION
**Pull request by hhyasdf**
Pull Request Description
---

### Describe what this PR does / why we need it
Support IP retain for Kubevirt VM

### Special notes for reviews

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -25,50 +24,120 @@ require (
...
+	k8s.io/client-go v0.23.6
+	k8s.io/component-base v0.23.6
 	k8s.io/kubernetes v0.0.0-00010101000000-000000000000
 	k8s.io/utils v0.0.0-20211208161948-7d6a63dca704
-	moul.io/http2curl v1.0.0 // indirect
+	kubevirt.io/api v0.54.0
 	sigs.k8s.io/controller-runtime v0.0.0-00010101000000-000000000000
 )
 
+require (
```
Comment (mars1024) [inline go.mod:40]: what's this require section for? seems all indirect dependencies

File: pkg/controllers/networking/ipinstance_controller.go [hunk 2]
```diff
@@ -57,7 +57,8 @@ func (r *IPInstanceReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	}
 
 	if !ip.DeletionTimestamp.IsZero() {
-		r.PodIPCache.Release(ip.Name, ip.Namespace)
+
```
[the remark this answers is not in the document]
Comment (hhyasdf) [reply pkg/controllers/networking/ipinstance_controller.go]: fixed

File: pkg/controllers/networking/pod_controller.go [hunk 3]
```diff
@@ -109,8 +117,57 @@ func (r *PodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result
...
+					return ctrl.Result{}, wrapError("unable to remove finalizer", r.removeFinalizer(ctx, pod))
+				}
+
+				log.V(1).Info("reserve ip for VM pod")
+				if err = r.reserve(ctx, pod, types.DropPodName(true)); err != nil {
+					return ctrl.Result{}, wrapError("unable to reserve pod", err)
+				}
+
+				vmIPInstances, err := utils.ListAllocatedIPInstances(ctx, r, client.MatchingLabels{
+					constants.LabelVM: vmName,
```
[the remark this answers is not in the document]
Comment (hhyasdf) [reply pkg/controllers/networking/pod_controller.go:145]: done

File: pkg/controllers/networking/pod_controller.go [hunk 4]
```diff
@@ -338,8 +405,48 @@ func (r *PodReconciler) statefulAllocate(ctx context.Context, pod *corev1.Pod, n
...
+	}
+
+	vmLabels := client.MatchingLabels{
+		constants.LabelVM: vmName,
+	}
+
+	var ipCandidates []ipCandidate
+	var allocatedIPInstances []*networkingv1.IPInstance
+
+	if allocatedIPInstances, err = utils.ListAllocatedIPInstances(ctx, r, vmLabels); err != nil {
```
[the remark this answers is not in the document]
Comment (hhyasdf) [reply pkg/controllers/networking/pod_controller.go]: done

**Review by mars1024 — APPROVED** [review]
/lgtm

**Pull request merged**